src/compiler/generate-common-tests.xsl

```diff
@@ -633,6 +645,46 @@
       </xsl:apply-templates>
    </xsl:template>
 
+   <!-- Generate error message for user-defined usage of names in XSpec namespace.
+        Context node is an x:variable element. -->
+   <xsl:template name="x:detect-reserved-variable-name" as="empty-sequence()">
+      <xsl:variable name="msg" as="xs:string"
+         select="'User-defined XSpec variables must not use the XSpec namespace.'"/>
+      <xsl:choose>
+         <xsl:when test="starts-with(normalize-space(@name),'Q{')">
+            <!-- URI-qualified name -->
+            <xsl:if test="replace(@name,'(^\s*Q\{)|(\}.*$)','') eq $xspec-namespace">
+               <xsl:sequence select="error(xs:QName('x:XSPEC008'), $msg)"/>
+            </xsl:if>
+         </xsl:when>
+         <xsl:when test="namespace-uri-from-QName(resolve-QName(@name,.)) eq $xspec-namespace">
+            <xsl:sequence select="error(xs:QName('x:XSPEC008'), $msg)"/>
+         </xsl:when>
+      </xsl:choose>
+   </xsl:template>
+
+   <!-- Given <vars> elements from tunnel parameter, return distinct EQNames.
+        The tunnel parameter originates in the match="x:variable" template with
+        mode="x:generate-calls" or mode="x:compile". -->
+   <xsl:function name="x:distinct-variable-names" as="xs:string*">
+      <xsl:param name="vars" as="element(var)*"/>
+      <!-- Create sequence of xs:QName values, so we can use distinct-values to compare them all. -->
+      <xsl:variable name="qnames" as="xs:QName*"
+         select="for $thisvar in $vars return
+         if (starts-with(normalize-space($thisvar/@name),'Q{'))
+         then QName(replace($thisvar/@name,'(^\s*Q\{)|(\}.*$)',''), substring-after($thisvar/@name,'}'))
+         else QName($thisvar/@namespace-uri, $thisvar/@name)
+         "/>
+      <xsl:variable name="distinctqnames" as="xs:QName*" select="distinct-values($qnames)"/>
+      <!-- Return distinct strings, except that any unprefixed name with nonempty URI
+         needs Q{} notation to record the namespace URI. -->
+      <xsl:sequence select="for $thisqname in $distinctqnames return
+         if ( empty(prefix-from-QName($thisqname)) and (string-length(namespace-uri-from-QName($thisqname)) gt 0) )
+         then 'Q{' || namespace-uri-from-QName($thisqname) || '}' || local-name-from-QName($thisqname)
```

Comment:
This breaks Saxon 9.7:

```console
E:\xspec\tutorial>java -cp "%SAXON_CP%" net.sf.saxon.Version
SAXON-HE 9.7.0.21J from Saxonica (build 072717)

E:\xspec\tutorial>..\bin\xspec.bat -t escape-for-regex.xspec
Saxon script not found, invoking JVM directly instead.

Creating Test Stylesheet...
Syntax error at char 216 near {...-from-QName($thisqname) || ...} in expression in xsl:sequence/@select on line 685 column 14 of generate-common-tests.xsl:
  XPST0003: Concatenation operator ('||') requires XPath 3.0 to be enabled
Errors were reported during stylesheet compilation

*** Error compiling the test suite
```

We will explicitly abandon ~9.7 in near future, as it is no longer maintained by Saxonica. But I'd like to keep compatibility for now.

Reply:
Thanks for noticing. I uploaded 78b43a5b2ce130a8ca535e1d12965cdfbddc9853. Switching from `||` to `concat()` was easy. I also had to find a home for some test code in `xspec-variable.xspec` that requires a later XPath version. I thought it was better to adapt and move the specific bits of code to `xspec-eqname.xspec` (which already has `xslt-version="3.0"`) rather than add `xslt-version="3.0"` to `xspec-variable.xspec`.

Reply:
I meant to add: I tested 78b43a5b2ce130a8ca535e1d12965cdfbddc9853 on my computer with Saxon-EE 9.7.0.8.